homeassistant/components/modem_callerid/__init__.py

```diff
@@ -1 +1,74 @@
-"""The modem_callerid component."""
+"""The Modem Caller ID integration."""
+import asyncio
+import logging
+
+from phone_modem import PhoneModem, exceptions
+import serial
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_DEVICE
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import DATA_KEY_API, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config):
+    """Set up the Modem Caller ID component."""
+    hass.data[DOMAIN] = {}
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config.get(DOMAIN)
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Modem Caller ID from a config entry."""
+    device = entry.data[CONF_DEVICE]
+    try:
+        api = PhoneModem(device)
```

Comment:
This needs to be wrapped in `hass.async_add_executor_job()` since creating `PhoneModem` will do I/O which will cause everything in Home Assistant to wait until it completes.

Ideally instantiating the `PhoneModem` class does not have any side effects (like doing I/O or connecting)

Then you can create the class in async context and then wrap your I/O calls to be run in an executor thread with something like `await hass.async_add_executor_job(api.connect)` which will run in https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor

Reply:
I am currently exploring `aioserial`. Attribute errors have me puzzled when starting off the library with `class PhoneModem(aioserial.AioSerial, serial.Serial):` `'PhoneModem' object has no attribute '_read_executor'`